flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java

```diff
@@ -414,6 +431,30 @@ private void maybeRegisterKafkaConsumerMetrics(
         }
     }
 
+    private void registerPendingRecordsMetric(
+            SourceReaderMetricGroup metricGroup, KafkaConsumer<?, ?> consumer) {
+
+        final String recordsLagMaxName = "records-lag-max";
```

Comment:
I wonder if this is the correct metric to track. AFAIK this metric describes the following 
`The maximum lag in terms of number of records for any partition in this window` so every subtask will show the same lag value. 

I know there is a similar metric like `records-lag` which is defined per partition. One could compute the sum of all the currently assigned partitions to get a proper lag calculation.

Unfortunately, the last time I looked into these metrics I remember vaguely that the _per-partition_ metrics are only registered after one record was consumed.

Reply:
Thanks for the careful review! This is really annoying... I'm not sure if that worth it to use additional data structures and maybe affect performance to implement just one metric. Is there any suggestions for ```pendingRecords```? Otherwise I prefer to skip it for now 😢

Reply:
I did another round of digging in Kafka and my initial assumption was not fully correct the `records-lag-max` metric is the maximum lag of all partitions assigned to the current consumer.
This gives us the possibility to simply multiply the current partition count by the lag value and we have a rough estimation.
WDYT?

Reply:
I think multiplying with #partition is a bit too inaccurate. If messages among subscribed partitions are imbalanced, the value of ```pendingRecords``` will be quite misleading to users.

Reply:
It can definitely be misleading when multiplying it in the case you've mentioned but on the other hand in the current situation, it is misleading if all assigned partitions lag equally behind. The `pendingRecords` would only account for one partition although the correct lag is up to partitions * `records-max-lag`. 

Personally, I can also think of not implementing this metric now and separate it from this PR. 
WDYT @AHeise?

Reply:
Honestly, you are thinking too hard. Simply expose `records-max-lag` as is. That's the metric that kafka exposes for it's consumer/fetcher, so that's exactly what Kafka users expect to see as a lag.